libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/NavigationRouteEngine.java

```diff
@@ -0,0 +1,203 @@
+   * <p>
+   * Uses {@link RouteOptions#coordinates()} and {@link RouteProgress#remainingWaypoints()}
+   * to determine the amount of remaining waypoints there are along the given route.
+   *
+   * @param location      current location of the device
+   * @param routeProgress for remaining waypoints along the route
+   * @since 0.13.0
+   */
+  @Override
+  public void findRouteFromRouteProgress(Location location, RouteProgress routeProgress) {
+    if (isValidProgress(location, routeProgress)) {
+      return;
+    }
+    this.routeProgress = routeProgress;
+    NavigationRoute.Builder builder = buildRouteRequest(location, routeProgress);
+    executeRouteCall(builder);
+  }
+
+  @Override
+  public void findRouteFromOriginToDestination(Point origin, Point destination) {
+    NavigationRoute.Builder builder = NavigationRoute.builder()
+      .accessToken(accessToken)
+      .origin(origin)
+      .destination(destination);
+    addLocaleAndUnitType(builder);
+    addRouteProfile(routeProfile, builder);
+    builder.build().getRoute(directionsResponseCallback);
+  }
+
+  @Nullable
+  private NavigationRoute.Builder buildRouteRequestFromCurrentLocation(Point origin, Double bearing,
+                                                                       RouteProgress progress,
+                                                                       @Nullable String routeProfile) {
+    RouteOptions options = progress.directionsRoute().routeOptions();
+    NavigationRoute.Builder builder = NavigationRoute.builder()
+      .origin(origin, bearing, BEARING_TOLERANCE)
+      .routeOptions(options);
+
+    List<Point> remainingWaypoints = RouteUtils.calculateRemainingWaypoints(progress);
+    if (remainingWaypoints == null) {
+      Timber.e("An error occurred fetching a new route");
+      return null;
+    }
+    addRouteProfile(routeProfile, builder);
+    addDestination(remainingWaypoints, builder);
+    addWaypoints(remainingWaypoints, builder);
+    return builder;
+  }
+
+  private void addLocaleAndUnitType(NavigationRoute.Builder builder) {
+    String voiceUnitType = NavigationUnitType.getDirectionsCriteriaUnitType(unitType, locale);
+    builder.language(locale).voiceUnits(voiceUnitType);
+  }
+
+  private void addRouteProfile(String routeProfile, NavigationRoute.Builder builder) {
+    if (!TextUtils.isEmpty(routeProfile)) {
+      builder.profile(routeProfile);
+    }
+  }
+
+  private void addDestination(List<Point> remainingWaypoints, NavigationRoute.Builder builder) {
+    if (!remainingWaypoints.isEmpty()) {
+      builder.destination(retrieveDestinationWaypoint(remainingWaypoints));
+    }
+  }
+
+  private Point retrieveDestinationWaypoint(List<Point> remainingWaypoints) {
+    int lastWaypoint = remainingWaypoints.size() - 1;
+    return remainingWaypoints.remove(lastWaypoint);
+  }
+
+  private static void addWaypoints(List<Point> remainingCoordinates, NavigationRoute.Builder builder) {
```

Comment:
👀 https://github.com/mapbox/mapbox-navigation-android/pull/894#discussion_r185135346

It seems you missed this one. Any reasons?